// @elastic/eui/test-env/components/header/header_alert/header_alert.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiHeaderAlert = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _flex = require("../../flex");

var _services = require("../../../services");

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var EuiHeaderAlert = function EuiHeaderAlert(_ref) {
  var action = _ref.action,
      className = _ref.className,
      date = _ref.date,
      text = _ref.text,
      title = _ref.title,
      badge = _ref.badge,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["action", "className", "date", "text", "title", "badge"]);
  var classes = (0, _classnames.default)('euiHeaderAlert', className);
  var ariaId = (0, _services.htmlIdGenerator)()();
  return _react.default.createElement("article", (0, _extends2.default)({
    "aria-labelledby": "".concat(ariaId, "-title"),
    className: classes
  }, rest), _react.default.createElement(_flex.EuiFlexGroup, {
    justifyContent: "spaceBetween"
  }, _react.default.createElement(_flex.EuiFlexItem, null, _react.default.createElement("div", {
    className: "euiHeaderAlert__date"
  }, date)), badge && _react.default.createElement(_flex.EuiFlexItem, {
    grow: false
  }, badge)), _react.default.createElement("h3", {
    id: "".concat(ariaId, "-title"),
    className: "euiHeaderAlert__title"
  }, title), _react.default.createElement("div", {
    className: "euiHeaderAlert__text"
  }, text), action && _react.default.createElement("div", {
    className: "euiHeaderAlert__action euiLink"
  }, action));
};

exports.EuiHeaderAlert = EuiHeaderAlert;
EuiHeaderAlert.propTypes = {
  className: _propTypes.default.string,
  "aria-label": _propTypes.default.string,
  "data-test-subj": _propTypes.default.string,

  /**
       * Adds a link to the alert.
       */
  action: _propTypes.default.node,
  date: _propTypes.default.node.isRequired,
  text: _propTypes.default.node,
  title: _propTypes.default.node.isRequired,

  /**
       * Accepts an `EuiBadge` that displays on the alert
       */
  badge: _propTypes.default.node
};
EuiHeaderAlert.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiHeaderAlert",
  "props": {
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "aria-label": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "data-test-subj": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "action": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": "Adds a link to the alert."
    },
    "date": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": ""
    },
    "text": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    },
    "title": {
      "type": {
        "name": "node"
      },
      "required": true,
      "description": ""
    },
    "badge": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": "Accepts an `EuiBadge` that displays on the alert"
    }
  }
};
