// @elastic/eui/lib/services/index.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "accessibleClickKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.accessibleClickKeys;
  }
});
Object.defineProperty(exports, "cascadingMenuKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.cascadingMenuKeys;
  }
});
Object.defineProperty(exports, "comboBoxKeys", {
  enumerable: true,
  get: function get() {
    return _accessibility.comboBoxKeys;
  }
});
Object.defineProperty(exports, "htmlIdGenerator", {
  enumerable: true,
  get: function get() {
    return _accessibility.htmlIdGenerator;
  }
});
Object.defineProperty(exports, "LEFT_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.LEFT_ALIGNMENT;
  }
});
Object.defineProperty(exports, "RIGHT_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.RIGHT_ALIGNMENT;
  }
});
Object.defineProperty(exports, "CENTER_ALIGNMENT", {
  enumerable: true,
  get: function get() {
    return _alignment.CENTER_ALIGNMENT;
  }
});
Object.defineProperty(exports, "isColorDark", {
  enumerable: true,
  get: function get() {
    return _color.isColorDark;
  }
});
Object.defineProperty(exports, "isValidHex", {
  enumerable: true,
  get: function get() {
    return _color.isValidHex;
  }
});
Object.defineProperty(exports, "calculateContrast", {
  enumerable: true,
  get: function get() {
    return _color.calculateContrast;
  }
});
Object.defineProperty(exports, "calculateLuminance", {
  enumerable: true,
  get: function get() {
    return _color.calculateLuminance;
  }
});
Object.defineProperty(exports, "hexToHsv", {
  enumerable: true,
  get: function get() {
    return _color.hexToHsv;
  }
});
Object.defineProperty(exports, "hexToRgb", {
  enumerable: true,
  get: function get() {
    return _color.hexToRgb;
  }
});
Object.defineProperty(exports, "hsvToHex", {
  enumerable: true,
  get: function get() {
    return _color.hsvToHex;
  }
});
Object.defineProperty(exports, "hsvToRgb", {
  enumerable: true,
  get: function get() {
    return _color.hsvToRgb;
  }
});
Object.defineProperty(exports, "rgbToHex", {
  enumerable: true,
  get: function get() {
    return _color.rgbToHex;
  }
});
Object.defineProperty(exports, "rgbToHsv", {
  enumerable: true,
  get: function get() {
    return _color.rgbToHsv;
  }
});
Object.defineProperty(exports, "VISUALIZATION_COLORS", {
  enumerable: true,
  get: function get() {
    return _color.VISUALIZATION_COLORS;
  }
});
Object.defineProperty(exports, "DEFAULT_VISUALIZATION_COLOR", {
  enumerable: true,
  get: function get() {
    return _color.DEFAULT_VISUALIZATION_COLOR;
  }
});
Object.defineProperty(exports, "colorPalette", {
  enumerable: true,
  get: function get() {
    return _color.colorPalette;
  }
});
Object.defineProperty(exports, "euiPaletteForLightBackground", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForLightBackground;
  }
});
Object.defineProperty(exports, "euiPaletteForDarkBackground", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForDarkBackground;
  }
});
Object.defineProperty(exports, "euiPaletteColorBlind", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteColorBlind;
  }
});
Object.defineProperty(exports, "euiPaletteColorBlindBehindText", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteColorBlindBehindText;
  }
});
Object.defineProperty(exports, "euiPaletteForStatus", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForStatus;
  }
});
Object.defineProperty(exports, "euiPaletteForTemperature", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteForTemperature;
  }
});
Object.defineProperty(exports, "euiPaletteComplimentary", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteComplimentary;
  }
});
Object.defineProperty(exports, "euiPaletteNegative", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteNegative;
  }
});
Object.defineProperty(exports, "euiPalettePositive", {
  enumerable: true,
  get: function get() {
    return _color.euiPalettePositive;
  }
});
Object.defineProperty(exports, "euiPaletteCool", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteCool;
  }
});
Object.defineProperty(exports, "euiPaletteWarm", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteWarm;
  }
});
Object.defineProperty(exports, "euiPaletteGray", {
  enumerable: true,
  get: function get() {
    return _color.euiPaletteGray;
  }
});
Object.defineProperty(exports, "useColorPickerState", {
  enumerable: true,
  get: function get() {
    return _color_picker.useColorPickerState;
  }
});
Object.defineProperty(exports, "useColorStopsState", {
  enumerable: true,
  get: function get() {
    return _color_picker.useColorStopsState;
  }
});
Object.defineProperty(exports, "copyToClipboard", {
  enumerable: true,
  get: function get() {
    return _copy_to_clipboard.copyToClipboard;
  }
});
Object.defineProperty(exports, "formatAuto", {
  enumerable: true,
  get: function get() {
    return _format.formatAuto;
  }
});
Object.defineProperty(exports, "formatBoolean", {
  enumerable: true,
  get: function get() {
    return _format.formatBoolean;
  }
});
Object.defineProperty(exports, "formatDate", {
  enumerable: true,
  get: function get() {
    return _format.formatDate;
  }
});
Object.defineProperty(exports, "formatNumber", {
  enumerable: true,
  get: function get() {
    return _format.formatNumber;
  }
});
Object.defineProperty(exports, "formatText", {
  enumerable: true,
  get: function get() {
    return _format.formatText;
  }
});
Object.defineProperty(exports, "dateFormatAliases", {
  enumerable: true,
  get: function get() {
    return _format.dateFormatAliases;
  }
});
Object.defineProperty(exports, "isEvenlyDivisibleBy", {
  enumerable: true,
  get: function get() {
    return _number.isEvenlyDivisibleBy;
  }
});
Object.defineProperty(exports, "isWithinRange", {
  enumerable: true,
  get: function get() {
    return _number.isWithinRange;
  }
});
Object.defineProperty(exports, "Pager", {
  enumerable: true,
  get: function get() {
    return _paging.Pager;
  }
});
Object.defineProperty(exports, "Random", {
  enumerable: true,
  get: function get() {
    return _random.Random;
  }
});
Object.defineProperty(exports, "getSecureRelForTarget", {
  enumerable: true,
  get: function get() {
    return _security.getSecureRelForTarget;
  }
});
Object.defineProperty(exports, "toInitials", {
  enumerable: true,
  get: function get() {
    return _string.toInitials;
  }
});
Object.defineProperty(exports, "PropertySortType", {
  enumerable: true,
  get: function get() {
    return _sort.PropertySortType;
  }
});
Object.defineProperty(exports, "SortDirectionType", {
  enumerable: true,
  get: function get() {
    return _sort.SortDirectionType;
  }
});
Object.defineProperty(exports, "SortDirection", {
  enumerable: true,
  get: function get() {
    return _sort.SortDirection;
  }
});
Object.defineProperty(exports, "SortableProperties", {
  enumerable: true,
  get: function get() {
    return _sort.SortableProperties;
  }
});
Object.defineProperty(exports, "Comparators", {
  enumerable: true,
  get: function get() {
    return _sort.Comparators;
  }
});
Object.defineProperty(exports, "calculatePopoverPosition", {
  enumerable: true,
  get: function get() {
    return _popover.calculatePopoverPosition;
  }
});
Object.defineProperty(exports, "findPopoverPosition", {
  enumerable: true,
  get: function get() {
    return _popover.findPopoverPosition;
  }
});
Object.defineProperty(exports, "getDurationAndPerformOnFrame", {
  enumerable: true,
  get: function get() {
    return _transition.getDurationAndPerformOnFrame;
  }
});
Object.defineProperty(exports, "getTransitionTimings", {
  enumerable: true,
  get: function get() {
    return _transition.getTransitionTimings;
  }
});
Object.defineProperty(exports, "getWaitDuration", {
  enumerable: true,
  get: function get() {
    return _transition.getWaitDuration;
  }
});
Object.defineProperty(exports, "performOnFrame", {
  enumerable: true,
  get: function get() {
    return _transition.performOnFrame;
  }
});
Object.defineProperty(exports, "EuiWindowEvent", {
  enumerable: true,
  get: function get() {
    return _window_event.EuiWindowEvent;
  }
});
Object.defineProperty(exports, "useDependentState", {
  enumerable: true,
  get: function get() {
    return _hooks.useDependentState;
  }
});
exports.keys = void 0;

var keys = _interopRequireWildcard(require("./keys"));

exports.keys = keys;

var _accessibility = require("./accessibility");

var _alignment = require("./alignment");

var _color = require("./color");

var _color_picker = require("./color_picker");

var _copy_to_clipboard = require("./copy_to_clipboard");

var _format = require("./format");

var _number = require("./number");

var _paging = require("./paging");

var _random = require("./random");

var _security = require("./security");

var _string = require("./string");

var _sort = require("./sort");

var _popover = require("./popover");

var _transition = require("./transition");

var _window_event = require("./window_event");

var _hooks = require("./hooks");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
