// @elastic/eui/es/components/search_bar/search_filters.js
function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { Component, Fragment } from 'react';
import PropTypes from "prop-types";
import { createFilter, SearchFilterConfig } from './filters';
import { EuiFilterGroup } from '../filter_group';
export {} from './filters';
export var EuiSearchFilters =
/*#__PURE__*/
function (_Component) {
  _inherits(EuiSearchFilters, _Component);

  function EuiSearchFilters() {
    _classCallCheck(this, EuiSearchFilters);

    return _possibleConstructorReturn(this, _getPrototypeOf(EuiSearchFilters).apply(this, arguments));
  }

  _createClass(EuiSearchFilters, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _this$props$filters = _this$props.filters,
          filters = _this$props$filters === void 0 ? [] : _this$props$filters,
          query = _this$props.query,
          onChange = _this$props.onChange;
      var items = [];
      filters.forEach(function (filterConfig, index) {
        if (filterConfig.available && !filterConfig.available()) {
          return;
        }

        var key = "filter_".concat(index);
        var control = createFilter(index, filterConfig, query, onChange);
        items.push(React.createElement(Fragment, {
          key: key
        }, control));
      });
      return React.createElement(EuiFilterGroup, null, items);
    }
  }]);

  return EuiSearchFilters;
}(Component);

_defineProperty(EuiSearchFilters, "defaultProps", {
  filters: []
});

EuiSearchFilters.propTypes = {
  query: PropTypes.any.isRequired,
  onChange: PropTypes.func.isRequired,
  filters: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.shape({
    type: PropTypes.oneOf(["is"]).isRequired,
    field: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    negatedName: PropTypes.string,
    available: PropTypes.func
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_selection"]).isRequired,
    field: PropTypes.string,
    name: PropTypes.string.isRequired,
    options: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape({
      field: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired, PropTypes.shape({
        type: PropTypes.oneOf(["date"]).isRequired,
        raw: PropTypes.any.isRequired,
        granularity: PropTypes.oneOfType([PropTypes.shape({
          es: PropTypes.oneOf(["d", "w", "M", "y"]).isRequired,
          js: PropTypes.oneOf(["day", "week", "month", "year"]).isRequired,
          isSame: PropTypes.func.isRequired,
          start: PropTypes.func.isRequired,
          startOfNext: PropTypes.func.isRequired,
          iso8601: PropTypes.func.isRequired
        }).isRequired, PropTypes.oneOf([undefined])]).isRequired,
        text: PropTypes.string.isRequired,
        resolve: PropTypes.func.isRequired
      }).isRequired]).isRequired,
      name: PropTypes.string,
      view: PropTypes.node
    }).isRequired).isRequired, PropTypes.func.isRequired]).isRequired,
    filterWith: PropTypes.oneOfType([PropTypes.oneOf(["prefix", "includes"]), PropTypes.func.isRequired]),
    cache: PropTypes.number,
    multiSelect: PropTypes.oneOfType([PropTypes.bool.isRequired, PropTypes.oneOf(["and", "or"])]),
    loadingMessage: PropTypes.string,
    noOptionsMessage: PropTypes.string,
    searchThreshold: PropTypes.number,
    available: PropTypes.func,
    autoClose: PropTypes.bool
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_toggle"]).isRequired,
    field: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired, PropTypes.shape({
      type: PropTypes.oneOf(["date"]).isRequired,
      raw: PropTypes.any.isRequired,
      granularity: PropTypes.oneOfType([PropTypes.shape({
        es: PropTypes.oneOf(["d", "w", "M", "y"]).isRequired,
        js: PropTypes.oneOf(["day", "week", "month", "year"]).isRequired,
        isSame: PropTypes.func.isRequired,
        start: PropTypes.func.isRequired,
        startOfNext: PropTypes.func.isRequired,
        iso8601: PropTypes.func.isRequired
      }).isRequired, PropTypes.oneOf([undefined])]).isRequired,
      text: PropTypes.string.isRequired,
      resolve: PropTypes.func.isRequired
    }).isRequired]).isRequired,
    name: PropTypes.string.isRequired,
    negatedName: PropTypes.string,
    available: PropTypes.func,
    operator: PropTypes.oneOf(["eq", "exact", "gt", "gte", "lt", "lte"])
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_toggle_group"]).isRequired,
    field: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired]).isRequired,
      name: PropTypes.string.isRequired,
      negatedName: PropTypes.string,
      operator: PropTypes.oneOf(["eq", "exact", "gt", "gte", "lt", "lte"])
    }).isRequired).isRequired,
    available: PropTypes.func
  }).isRequired]).isRequired).isRequired
};
EuiSearchFilters.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiSearchFilters",
  "props": {
    "filters": {
      "defaultValue": {
        "value": "[]",
        "computed": false
      },
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "union",
          "value": [{
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"is\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "name": {
                "name": "string",
                "required": true
              },
              "negatedName": {
                "name": "string",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_selection\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": false
              },
              "name": {
                "name": "string",
                "required": true
              },
              "options": {
                "name": "union",
                "value": [{
                  "name": "arrayOf",
                  "value": {
                    "name": "shape",
                    "value": {
                      "field": {
                        "name": "string",
                        "required": false
                      },
                      "value": {
                        "name": "union",
                        "value": [{
                          "name": "string"
                        }, {
                          "name": "number"
                        }, {
                          "name": "bool"
                        }, {
                          "name": "shape",
                          "value": {
                            "type": {
                              "name": "enum",
                              "value": [{
                                "value": "\"date\"",
                                "computed": false
                              }],
                              "required": true
                            },
                            "raw": {
                              "name": "any",
                              "required": true
                            },
                            "granularity": {
                              "name": "union",
                              "value": [{
                                "name": "shape",
                                "value": {
                                  "es": {
                                    "name": "enum",
                                    "value": [{
                                      "value": "\"d\"",
                                      "computed": false
                                    }, {
                                      "value": "\"w\"",
                                      "computed": false
                                    }, {
                                      "value": "\"M\"",
                                      "computed": false
                                    }, {
                                      "value": "\"y\"",
                                      "computed": false
                                    }],
                                    "required": true
                                  },
                                  "js": {
                                    "name": "enum",
                                    "value": [{
                                      "value": "\"day\"",
                                      "computed": false
                                    }, {
                                      "value": "\"week\"",
                                      "computed": false
                                    }, {
                                      "value": "\"month\"",
                                      "computed": false
                                    }, {
                                      "value": "\"year\"",
                                      "computed": false
                                    }],
                                    "required": true
                                  },
                                  "isSame": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "start": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "startOfNext": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "iso8601": {
                                    "name": "func",
                                    "required": true
                                  }
                                }
                              }, {
                                "name": "enum",
                                "value": [{
                                  "value": "undefined",
                                  "computed": true
                                }]
                              }],
                              "required": true
                            },
                            "text": {
                              "name": "string",
                              "required": true
                            },
                            "resolve": {
                              "name": "func",
                              "required": true
                            }
                          }
                        }],
                        "required": true
                      },
                      "name": {
                        "name": "string",
                        "required": false
                      },
                      "view": {
                        "name": "node",
                        "required": false
                      }
                    }
                  }
                }, {
                  "name": "func"
                }],
                "required": true
              },
              "filterWith": {
                "name": "union",
                "value": [{
                  "name": "enum",
                  "value": [{
                    "value": "\"prefix\"",
                    "computed": false
                  }, {
                    "value": "\"includes\"",
                    "computed": false
                  }]
                }, {
                  "name": "func"
                }],
                "required": false
              },
              "cache": {
                "name": "number",
                "required": false
              },
              "multiSelect": {
                "name": "union",
                "value": [{
                  "name": "bool"
                }, {
                  "name": "enum",
                  "value": [{
                    "value": "\"and\"",
                    "computed": false
                  }, {
                    "value": "\"or\"",
                    "computed": false
                  }]
                }],
                "required": false
              },
              "loadingMessage": {
                "name": "string",
                "required": false
              },
              "noOptionsMessage": {
                "name": "string",
                "required": false
              },
              "searchThreshold": {
                "name": "number",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              },
              "autoClose": {
                "name": "bool",
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_toggle\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "value": {
                "name": "union",
                "value": [{
                  "name": "string"
                }, {
                  "name": "number"
                }, {
                  "name": "bool"
                }, {
                  "name": "shape",
                  "value": {
                    "type": {
                      "name": "enum",
                      "value": [{
                        "value": "\"date\"",
                        "computed": false
                      }],
                      "required": true
                    },
                    "raw": {
                      "name": "any",
                      "required": true
                    },
                    "granularity": {
                      "name": "union",
                      "value": [{
                        "name": "shape",
                        "value": {
                          "es": {
                            "name": "enum",
                            "value": [{
                              "value": "\"d\"",
                              "computed": false
                            }, {
                              "value": "\"w\"",
                              "computed": false
                            }, {
                              "value": "\"M\"",
                              "computed": false
                            }, {
                              "value": "\"y\"",
                              "computed": false
                            }],
                            "required": true
                          },
                          "js": {
                            "name": "enum",
                            "value": [{
                              "value": "\"day\"",
                              "computed": false
                            }, {
                              "value": "\"week\"",
                              "computed": false
                            }, {
                              "value": "\"month\"",
                              "computed": false
                            }, {
                              "value": "\"year\"",
                              "computed": false
                            }],
                            "required": true
                          },
                          "isSame": {
                            "name": "func",
                            "required": true
                          },
                          "start": {
                            "name": "func",
                            "required": true
                          },
                          "startOfNext": {
                            "name": "func",
                            "required": true
                          },
                          "iso8601": {
                            "name": "func",
                            "required": true
                          }
                        }
                      }, {
                        "name": "enum",
                        "value": [{
                          "value": "undefined",
                          "computed": true
                        }]
                      }],
                      "required": true
                    },
                    "text": {
                      "name": "string",
                      "required": true
                    },
                    "resolve": {
                      "name": "func",
                      "required": true
                    }
                  }
                }],
                "required": true
              },
              "name": {
                "name": "string",
                "required": true
              },
              "negatedName": {
                "name": "string",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              },
              "operator": {
                "name": "enum",
                "value": [{
                  "value": "\"eq\"",
                  "computed": false
                }, {
                  "value": "\"exact\"",
                  "computed": false
                }, {
                  "value": "\"gt\"",
                  "computed": false
                }, {
                  "value": "\"gte\"",
                  "computed": false
                }, {
                  "value": "\"lt\"",
                  "computed": false
                }, {
                  "value": "\"lte\"",
                  "computed": false
                }],
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_toggle_group\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "items": {
                "name": "arrayOf",
                "value": {
                  "name": "shape",
                  "value": {
                    "value": {
                      "name": "union",
                      "value": [{
                        "name": "string"
                      }, {
                        "name": "number"
                      }, {
                        "name": "bool"
                      }],
                      "required": true
                    },
                    "name": {
                      "name": "string",
                      "required": true
                    },
                    "negatedName": {
                      "name": "string",
                      "required": false
                    },
                    "operator": {
                      "name": "enum",
                      "value": [{
                        "value": "\"eq\"",
                        "computed": false
                      }, {
                        "value": "\"exact\"",
                        "computed": false
                      }, {
                        "value": "\"gt\"",
                        "computed": false
                      }, {
                        "value": "\"gte\"",
                        "computed": false
                      }, {
                        "value": "\"lt\"",
                        "computed": false
                      }, {
                        "value": "\"lte\"",
                        "computed": false
                      }],
                      "required": false
                    }
                  }
                },
                "required": true
              },
              "available": {
                "name": "func",
                "required": false
              }
            }
          }]
        }
      },
      "required": false,
      "description": ""
    },
    "query": {
      "type": {
        "name": "any"
      },
      "required": true,
      "description": ""
    },
    "onChange": {
      "type": {
        "name": "func"
      },
      "required": true,
      "description": ""
    }
  }
};
