// @elastic/eui/test-env/components/selectable/selectable_list/selectable_list.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiSelectableList = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _reactVirtualized = require("react-virtualized");

var _services = require("../../../services");

var _selectable_list_item = require("./selectable_list_item");

var _highlight = require("../../highlight");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var EuiSelectableList =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(EuiSelectableList, _Component);

  function EuiSelectableList(props) {
    var _this;

    (0, _classCallCheck2.default)(this, EuiSelectableList);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(EuiSelectableList).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "rootId", _this.props.rootId || (0, _services.htmlIdGenerator)());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onAddOrRemoveOption", function (option) {
      if (option.disabled) {
        return;
      }

      var allowExclusions = _this.props.allowExclusions;

      if (option.checked === 'on' && allowExclusions) {
        _this.onExcludeOption(option);
      } else if (option.checked === 'on' || option.checked === 'off') {
        _this.onRemoveOption(option);
      } else {
        _this.onAddOption(option);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onAddOption", function (addedOption) {
      var _this$props = _this.props,
          onOptionClick = _this$props.onOptionClick,
          options = _this$props.options,
          singleSelection = _this$props.singleSelection;
      var updatedOptions = options.map(function (option) {
        // if singleSelection is enabled, uncheck any selected option(s)
        var updatedOption = _objectSpread({}, option);

        if (singleSelection) {
          delete updatedOption.checked;
        } // if this is the now-selected option, check it


        if (option === addedOption) {
          updatedOption.checked = 'on';
        }

        return updatedOption;
      });
      onOptionClick(updatedOptions);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onRemoveOption", function (removedOption) {
      var _this$props2 = _this.props,
          onOptionClick = _this$props2.onOptionClick,
          singleSelection = _this$props2.singleSelection,
          options = _this$props2.options;
      var updatedOptions = options.map(function (option) {
        var updatedOption = _objectSpread({}, option);

        if (option === removedOption && singleSelection !== 'always') {
          delete updatedOption.checked;
        }

        return updatedOption;
      });
      onOptionClick(updatedOptions);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onExcludeOption", function (excludedOption) {
      var _this$props3 = _this.props,
          onOptionClick = _this$props3.onOptionClick,
          options = _this$props3.options;
      excludedOption.checked = 'off';
      var updatedOptions = options.map(function (option) {
        var updatedOption = _objectSpread({}, option);

        if (option === excludedOption) {
          updatedOption.checked = 'off';
        }

        return updatedOption;
      });
      onOptionClick(updatedOptions);
    });
    return _this;
  }

  (0, _createClass2.default)(EuiSelectableList, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props4 = this.props,
          className = _this$props4.className,
          options = _this$props4.options,
          searchValue = _this$props4.searchValue,
          onOptionClick = _this$props4.onOptionClick,
          renderOption = _this$props4.renderOption,
          forcedHeight = _this$props4.height,
          virtualizedProps = _this$props4.virtualizedProps,
          rowHeight = _this$props4.rowHeight,
          activeOptionIndex = _this$props4.activeOptionIndex,
          rootId = _this$props4.rootId,
          showIcons = _this$props4.showIcons,
          singleSelection = _this$props4.singleSelection,
          visibleOptions = _this$props4.visibleOptions,
          allowExclusions = _this$props4.allowExclusions,
          bordered = _this$props4.bordered,
          rest = (0, _objectWithoutProperties2.default)(_this$props4, ["className", "options", "searchValue", "onOptionClick", "renderOption", "height", "virtualizedProps", "rowHeight", "activeOptionIndex", "rootId", "showIcons", "singleSelection", "visibleOptions", "allowExclusions", "bordered"]);
      var optionArray = visibleOptions || options;
      var heightIsFull = forcedHeight === 'full';
      var calculatedHeight = heightIsFull ? false : forcedHeight; // If calculatedHeight is still undefined, then calculate it

      if (calculatedHeight === undefined) {
        var maxVisibleOptions = 7;
        var numVisibleOptions = optionArray.length;
        var numVisibleMoreThanMax = optionArray.length > maxVisibleOptions;

        if (numVisibleMoreThanMax) {
          // Show only half of the last one to indicate there's more to scroll to
          calculatedHeight = (maxVisibleOptions - 0.5) * rowHeight;
        } else {
          calculatedHeight = numVisibleOptions * rowHeight;
        }
      }

      var classes = (0, _classnames.default)('euiSelectableList', {
        'euiSelectableList-fullHeight': heightIsFull,
        'euiSelectableList-bordered': bordered
      }, className);
      return _react.default.createElement("div", (0, _extends2.default)({
        className: classes
      }, rest), _react.default.createElement(_reactVirtualized.AutoSizer, {
        disableHeight: !heightIsFull
      }, function (_ref) {
        var width = _ref.width,
            height = _ref.height;
        return _react.default.createElement(_reactVirtualized.List, (0, _extends2.default)({
          id: _this2.rootId('listbox'),
          className: "euiSelectableList__list",
          role: "listbox",
          width: width,
          height: calculatedHeight || height,
          rowCount: optionArray.length,
          rowHeight: rowHeight,
          scrollToIndex: activeOptionIndex
        }, virtualizedProps, {
          rowRenderer: function rowRenderer(_ref2) {
            var rowKey = _ref2.key,
                index = _ref2.index,
                style = _ref2.style;
            var option = optionArray[index];
            var label = option.label,
                isGroupLabel = option.isGroupLabel,
                checked = option.checked,
                disabled = option.disabled,
                prepend = option.prepend,
                append = option.append,
                ref = option.ref,
                key = option.key,
                optionRest = (0, _objectWithoutProperties2.default)(option, ["label", "isGroupLabel", "checked", "disabled", "prepend", "append", "ref", "key"]);

            if (isGroupLabel) {
              return _react.default.createElement("div", (0, _extends2.default)({
                className: "euiSelectableList__groupLabel",
                key: rowKey,
                style: style
              }, optionRest), prepend, label, append);
            }

            return _react.default.createElement(_selectable_list_item.EuiSelectableListItem, (0, _extends2.default)({
              id: _this2.rootId("_option-".concat(index)),
              style: style,
              key: key || option.label.toLowerCase(),
              onClick: function onClick() {
                return _this2.onAddOrRemoveOption(option);
              },
              ref: ref ? ref.bind(null, index) : undefined,
              isFocused: activeOptionIndex === index,
              title: label,
              showIcons: showIcons,
              checked: checked,
              disabled: disabled,
              prepend: prepend,
              append: append
            }, optionRest), renderOption ? renderOption(option, searchValue) : _react.default.createElement(_highlight.EuiHighlight, {
              search: searchValue
            }, label));
          }
        }));
      }));
    }
  }]);
  return EuiSelectableList;
}(_react.Component);

exports.EuiSelectableList = EuiSelectableList;
(0, _defineProperty2.default)(EuiSelectableList, "defaultProps", {
  rowHeight: 32,
  searchValue: ''
});
EuiSelectableList.propTypes = {
  className: _propTypes.default.string,
  "aria-label": _propTypes.default.string,
  "data-test-subj": _propTypes.default.string,

  /**
       * The index of the option to be highlighted as pseudo-focused;
       * Good for use when only one selection is allowed and needing to open
       * directly to that option
       */
  activeOptionIndex: _propTypes.default.number,

  /**
       *  The height of each option in pixels. Defaults to `32`
       */
  rowHeight: _propTypes.default.number.isRequired,

  /**
       * Show the check/cross selection indicator icons
       */
  showIcons: _propTypes.default.bool,
  singleSelection: _propTypes.default.oneOfType([_propTypes.default.oneOf(["always"]), _propTypes.default.bool.isRequired]),

  /**
       * Any props to send specifically to the react-virtualized `List`
       */
  virtualizedProps: _propTypes.default.any,

  /**
       * Adds a border around the list to indicate the bounds;
       * Useful when the list scrolls, otherwise use your own container
       */
  bordered: _propTypes.default.bool,

  /**
     * All possible options
     */
  options: _propTypes.default.arrayOf(_propTypes.default.shape({
    /**
       * Optional `boolean`.
       * Set to `true` to indicate object is just a grouping label, not a selectable item
       */
    isGroupLabel: _propTypes.default.oneOfType([_propTypes.default.oneOf([true]).isRequired, _propTypes.default.oneOf([false])]),

    /**
       * Visible label of option.
       * Must be unique across items if `key` is not supplied
       */
    label: _propTypes.default.string,

    /**
       * Must be unique across items.
       * Will be used to match options instead of `label`
       */
    key: _propTypes.default.string,

    /**
       * Leave `undefined` to indicate not selected,
       * 'on' to indicate inclusion and
       * 'off' to indicate exclusion
       */
    checked: _propTypes.default.oneOf(["on", "off", undefined]),
    disabled: _propTypes.default.bool,

    /**
       * Node to add between the selection icon and the label
       */
    prepend: _propTypes.default.node,

    /**
       * Node to add to the far right of the item
       */
    append: _propTypes.default.node,
    ref: _propTypes.default.func,
    className: _propTypes.default.string,
    "aria-label": _propTypes.default.string,
    "data-test-subj": _propTypes.default.string
  }).isRequired).isRequired,

  /**
     * Filtered options list (if applicable)
     */
  visibleOptions: _propTypes.default.arrayOf(_propTypes.default.shape({
    isGroupLabel: _propTypes.default.oneOfType([_propTypes.default.oneOf([true]).isRequired, _propTypes.default.oneOf([false])]),
    label: _propTypes.default.string,
    key: _propTypes.default.string,
    checked: _propTypes.default.oneOf(["on", "off", undefined]),
    disabled: _propTypes.default.bool,
    prepend: _propTypes.default.node,
    append: _propTypes.default.node,
    ref: _propTypes.default.func,
    className: _propTypes.default.string,
    "aria-label": _propTypes.default.string,
    "data-test-subj": _propTypes.default.string
  }).isRequired),

  /**
     * Search value to highlight on the option render
     */
  searchValue: _propTypes.default.string.isRequired,

  /**
     * Returns the array of options with altered checked state
     */
  onOptionClick: _propTypes.default.func.isRequired,

  /**
     * Custom render for the label portion of the option;
     * Takes (option, searchValue), returns ReactNode
     */
  renderOption: _propTypes.default.func,

  /**
     * Sets the max height in pixels or pass `full` to allow
     * the whole group to fill the height of its container and
     * allows the list grow as well
     */
  height: _propTypes.default.oneOfType([_propTypes.default.number.isRequired, _propTypes.default.oneOf(["full"])]),

  /**
     * Allow cycling through the on, off and undefined state of option.checked
     * and not just on and undefined
     */
  allowExclusions: _propTypes.default.bool,
  rootId: _propTypes.default.func
};
EuiSelectableList.__docgenInfo = {
  "description": "",
  "methods": [{
    "name": "onAddOrRemoveOption",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "option",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onAddOption",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "addedOption",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onRemoveOption",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "removedOption",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onExcludeOption",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "excludedOption",
      "type": null
    }],
    "returns": null
  }],
  "displayName": "EuiSelectableList",
  "props": {
    "rowHeight": {
      "defaultValue": {
        "value": "32",
        "computed": false
      },
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "The height of each option in pixels. Defaults to `32`"
    },
    "searchValue": {
      "defaultValue": {
        "value": "''",
        "computed": false
      },
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "Search value to highlight on the option render"
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "aria-label": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "data-test-subj": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "activeOptionIndex": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "The index of the option to be highlighted as pseudo-focused;\nGood for use when only one selection is allowed and needing to open\ndirectly to that option"
    },
    "showIcons": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Show the check/cross selection indicator icons"
    },
    "singleSelection": {
      "type": {
        "name": "union",
        "value": [{
          "name": "enum",
          "value": [{
            "value": "\"always\"",
            "computed": false
          }]
        }, {
          "name": "bool"
        }]
      },
      "required": false,
      "description": ""
    },
    "virtualizedProps": {
      "type": {
        "name": "any"
      },
      "required": false,
      "description": "Any props to send specifically to the react-virtualized `List`"
    },
    "bordered": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Adds a border around the list to indicate the bounds;\nUseful when the list scrolls, otherwise use your own container"
    },
    "options": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "shape",
          "value": {
            "isGroupLabel": {
              "name": "union",
              "value": [{
                "name": "enum",
                "value": [{
                  "value": "true",
                  "computed": false
                }]
              }, {
                "name": "enum",
                "value": [{
                  "value": "false",
                  "computed": false
                }]
              }],
              "description": "Optional `boolean`.\nSet to `true` to indicate object is just a grouping label, not a selectable item",
              "required": false
            },
            "label": {
              "name": "string",
              "description": "Visible label of option.\nMust be unique across items if `key` is not supplied",
              "required": false
            },
            "key": {
              "name": "string",
              "description": "Must be unique across items.\nWill be used to match options instead of `label`",
              "required": false
            },
            "checked": {
              "name": "enum",
              "value": [{
                "value": "\"on\"",
                "computed": false
              }, {
                "value": "\"off\"",
                "computed": false
              }, {
                "value": "undefined",
                "computed": true
              }],
              "description": "Leave `undefined` to indicate not selected,\n'on' to indicate inclusion and\n'off' to indicate exclusion",
              "required": false
            },
            "disabled": {
              "name": "bool",
              "required": false
            },
            "prepend": {
              "name": "node",
              "description": "Node to add between the selection icon and the label",
              "required": false
            },
            "append": {
              "name": "node",
              "description": "Node to add to the far right of the item",
              "required": false
            },
            "ref": {
              "name": "func",
              "required": false
            },
            "className": {
              "name": "string",
              "required": false
            },
            "aria-label": {
              "name": "string",
              "required": false
            },
            "data-test-subj": {
              "name": "string",
              "required": false
            }
          }
        }
      },
      "required": true,
      "description": "All possible options"
    },
    "visibleOptions": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "shape",
          "value": {
            "isGroupLabel": {
              "name": "union",
              "value": [{
                "name": "enum",
                "value": [{
                  "value": "true",
                  "computed": false
                }]
              }, {
                "name": "enum",
                "value": [{
                  "value": "false",
                  "computed": false
                }]
              }],
              "required": false
            },
            "label": {
              "name": "string",
              "required": false
            },
            "key": {
              "name": "string",
              "required": false
            },
            "checked": {
              "name": "enum",
              "value": [{
                "value": "\"on\"",
                "computed": false
              }, {
                "value": "\"off\"",
                "computed": false
              }, {
                "value": "undefined",
                "computed": true
              }],
              "required": false
            },
            "disabled": {
              "name": "bool",
              "required": false
            },
            "prepend": {
              "name": "node",
              "required": false
            },
            "append": {
              "name": "node",
              "required": false
            },
            "ref": {
              "name": "func",
              "required": false
            },
            "className": {
              "name": "string",
              "required": false
            },
            "aria-label": {
              "name": "string",
              "required": false
            },
            "data-test-subj": {
              "name": "string",
              "required": false
            }
          }
        }
      },
      "required": false,
      "description": "Filtered options list (if applicable)"
    },
    "onOptionClick": {
      "type": {
        "name": "func"
      },
      "required": true,
      "description": "Returns the array of options with altered checked state"
    },
    "renderOption": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": "Custom render for the label portion of the option;\nTakes (option, searchValue), returns ReactNode"
    },
    "height": {
      "type": {
        "name": "union",
        "value": [{
          "name": "number"
        }, {
          "name": "enum",
          "value": [{
            "value": "\"full\"",
            "computed": false
          }]
        }]
      },
      "required": false,
      "description": "Sets the max height in pixels or pass `full` to allow\nthe whole group to fill the height of its container and\nallows the list grow as well"
    },
    "allowExclusions": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Allow cycling through the on, off and undefined state of option.checked\nand not just on and undefined"
    },
    "rootId": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": ""
    }
  }
};
