// react-axios/es/components/withAxios.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Request from './Request';

export var withAxios = function withAxios() {
  var mixed = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  if (typeof mixed === 'function') {
    // basic axios provider HoC
    var WrappedComponent = mixed;
    var AxiosExtracter = function AxiosExtracter(props, context) {
      return React.createElement(WrappedComponent, _extends({ axios: context.axios || axios }, props));
    };

    AxiosExtracter.contextTypes = {
      axios: PropTypes.func
    };
    return AxiosExtracter;
  }
  // advanced Request provider HoC
  var options = _extends({}, Request.defaultProps, mixed);
  return function (WrappedComponent) {
    // validate the options passed in are valid request propTypes.
    PropTypes.checkPropTypes(_extends({}, Request.propTypes, {
      method: PropTypes.string // not required if the user just wants access to the axios instance
    }), options, 'option', 'withAxios()(' + WrappedComponent.name + ')');

    var ReactAxiosExtracter = function (_React$PureComponent) {
      _inherits(ReactAxiosExtracter, _React$PureComponent);

      function ReactAxiosExtracter() {
        _classCallCheck(this, ReactAxiosExtracter);

        return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
      }

      ReactAxiosExtracter.prototype.render = function render() {
        var _this2 = this;

        // allow overriding the config initial options
        var newOptions = _extends({}, options, this.props.options);
        return React.createElement(
          Request,
          newOptions,
          function (error, response, isLoading, makeRequest, axios) {
            return React.createElement(WrappedComponent, _extends({}, _this2.props, {
              error: error,
              response: response,
              isLoading: isLoading,
              makeRequest: makeRequest,
              axios: axios,
              options: newOptions
            }));
          }
        );
      };

      return ReactAxiosExtracter;
    }(React.PureComponent);

    ReactAxiosExtracter.propTypes = {
      options: PropTypes.object
    };
    return ReactAxiosExtracter;
  };
};

export default withAxios;
